### PR TITLE
holy water refactor + cult ghost bugfixing

### PR DESCRIPTION
**Pull request by ATH1909**
## About The Pull Request

Holy water now uses current_cycle instead of its own redundant data["misc"] thing. As a side effect of this, you technically require 0.4u less holy water to deconvert a cultist now, but that happens to end up giving us a nice, round number: 24u (unless you abuse the fact that the last tick doesn't need >=metab_rate of holy water remaining to work and use ~23.7u of holy water instead, but that's overcomplicating things).

Holy water no longer brings your jitteriness value down to 10 if it was already above 10.

Seizures caused by holy water can now only happen once per deconversion attempt, paralyze people for 6 seconds instead of knocking them unconscious for 12 seconds, cause extra jittering, can no longer occur on the same tick that holy water deconverts you on, and are more likely to occur than they were before. Nar'Sie now also has an expanded (and updated) repertoire of lines to roast you with when you have a holy water-induced seizure.

Holy water now removes itself from your stomach as well as your bloodstream when it reaches cycle 60, instead of only from your bloodstream.

Being a cult ghost (from a Spirit Realm cult rune) has been changed to be a trait instead of a status effect, since the status effect was bugged and wouldn't properly apply itself (and its effects could be easily replicated by a trait and some extra lines in the code for Spirit Realm runes).

The above bug fix also fixed a bug that allowed cult ghosts to activate spirit realm runes despite the code specifically saying that they were't supposed to be able to do that. Because of this, cult ghosts can no longer summon more cult ghosts using spirit realm runes. I decided to throw them a bone and allow them to still ascend as dark spirits, though.

Also because of the above bug fix, cult ghosts can now no longer be deconverted by holy water. In addition, I took things a bit further, and instead of repeatedly having all holy water purged from their systems (which was one of the things that the cult ghost status effect was supposed to do), cult ghosts from the manifest ghost rune are now harmed by exposure to holy water and by holy water ticking in their systems.

Spirit Realm runes now account for the existence of non-binary cult ghosts.

## Why It's Good For The Game

The code for holy water is painfully outdated.

image

I had been meaning to get around to fixing the "cult ghosts can summon more cult ghosts" bug for a while and mothblocks reminded me to go do it, so three hours later, here we are.

This should reduce the insanity of Spirit Realm rune+pylon stacks considerably, as well as prevent a single remaining (living) cultist from summoning Nar'Sie by just using a Spirit Realm rune and having the ghost cultists from their Spirit Realm rune summon enough ghost cultists to summon Nar'Sie.

I like The Wizard of Oz.

## Changelog
:cl: ATHATH
fix: Cult ghosts from Spirit Realm runes can no longer summon even more cult ghosts. They can still ascend as dark spirits, however.
fix: Cult ghosts can no longer be deconverted by holy water.
balance: Cult ghosts now take significant amounts of damage from being exposed to holy water and from having holy water in their systems.
fix: Spirit Realm runes now account for the existence of non-binary cult ghosts.
tweak: Seizures caused by holy water can now only happen once per deconversion attempt, paralyze people for 6 seconds instead of knocking them unconscious for 12 seconds, cause extra jittering, can no longer occur on the same tick that holy water deconverts you on, and are more likely to occur than they were before.
add: Nar'Sie now also has an expanded (and updated) repertoire of lines to roast you with when you have a holy water-induced seizure.
fix: Holy water now removes itself from your stomach as well as your bloodstream when it reaches cycle 60, instead of only from your bloodstream.
/:cl:

### CODE REVIEW

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 1]
```diff
@@ -218,54 +218,73 @@
...
 	if(iscultist(M))
 		for(var/datum/action/innate/cult/blood_magic/BM in M.actions)
 			to_chat(M, "<span class='cultlarge'>Your blood rites falter as holy water scours your body!</span>")
 			for(var/datum/action/innate/cult/blood_spell/BS in BM.spells)
 				qdel(BS)
-	if(data["misc"] >= 25)		// 10 units, 45 seconds @ metabolism 0.4 units & tick rate 1.8 sec
+
+	if(HAS_TRAIT(M, TRAIT_CULT_GHOST)) //it would be really awkward if summoned cult ghosts could be deconverted
+		exposed_mob.adjustFireLoss(5, 0) //and it burns, burns, burns~
+		return //note to self for a future PR: maybe add some heavy damage over time here, since most of a cult ghost's body is made of nar'sie's magic?
```
Comment (Fikou) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:258]: who is self

Comment (ATH1909) [reply code/modules/reagents/chemistry/reagents/other_reagents.dm]: the man behind the slaughter

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 2]
```diff
@@ -218,54 +218,73 @@
...
 	. = ..()
-	if(iscultist(exposed_mob))
-		to_chat(exposed_mob, "<span class='userdanger'>A vile holiness begins to spread its shining tendrils through your mind, purging the Geometer of Blood's influence!</span>")
+	if(HAS_TRAIT(exposed_mob, TRAIT_CULT_GHOST))
+		exposed_mob.adjustFireLoss(reac_volume, 0) //a bit extreme, perhaps, but cult ghosts are meant to be expendable and their bodies are MADE of unholy magic
+		to_chat(exposed_mob, "<span class='userdanger'>IT BURNS! IT BUUUUUURNS!</span>")
+		if((methods & TOUCH) && reac_volume >= 70 && exposed_mob.health <= exposed_mob.crit_threshold) //yo holy shit he dead
+			exposed_mob.say("AAAHHH! YOU CURSED BRAT! LOOK WHAT YOU'VE DONE! I'm MELTING! Melting! Ohhh, what a world, what a world! Who would have thought a good little spaceman like you could destroy my beautiful wickedness?! AUUUUGH! I'm gone! I'm gone! I'm going... Augh... Aaauuggghhh...", forced = "holy water") //you can theoretically say this and live using something that lets you avoid falling into crit, but eh, I'm not gonna bother to add a bunch of code for those edge cases for such a niche occurrence
```
Comment (necromanceranne) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:242]: Maybe just shorten this to the second 'I'm melting!' or else it might get very...obstructive.

Comment (ATH1909) [reply code/modules/reagents/chemistry/reagents/other_reagents.dm:242]: I could make it be a 10% chance to say the full thing and a 90% (mutually exclusive) chance to say the shortened version

Comment (coiax) [reply code/modules/reagents/chemistry/reagents/other_reagents.dm:242]: What would it take for you to get a reac volume of 70u though?

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 3]
```diff
@@ -218,54 +218,73 @@
...
-	if(M.blood_volume)
-		M.blood_volume += 0.1 // water is good for you!
-	if(!data)
-		data = list("misc" = 1)
-	data["misc"]++
-	M.jitteriness = min(M.jitteriness+4,10)
+	. = ..() //includes the increased blood (re)generation rate from normal water
+
+	if(M.jitteriness < 10)
+		M.jitteriness = min(M.jitteriness+4,10)
```
Comment (coiax) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:248]:
```suggestion
		M.jitteriness = min(M.jitteriness + 4, 10)
```

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 4]
```diff
@@ -218,54 +218,73 @@
...
+			if(prob(20) && !seizure) //4% chance/tick for Nar'Sie to roast you
 				M.visible_message("<span class='danger'>[M] starts having a seizure!</span>", "<span class='userdanger'>You have a seizure!</span>")
-				M.Unconscious(120)
-				to_chat(M, "<span class='cultlarge'>[pick("Your blood is your bond - you are nothing without it", "Do not forget your place", \
-				"All that power, and you still fail?", "If you cannot scour this poison, I shall scour your meager life!")].</span>")
-	if(data["misc"] >= 60)	// 30 units, 135 seconds
+				seizure = TRUE
+				M.Paralyze(60)
+				if(M.jitteriness < 40)
+					M.jitteriness = min(M.jitteriness+25,40)
```
Comment (coiax) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:272]:
```suggestion
					M.jitteriness = min(M.jitteriness + 25, 40)
```

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 5]
```diff
@@ -218,54 +218,73 @@
...
+				to_chat(M, "<span class='cultlarge'>[pick("Your blood is your bond - you are NOTHING without it.", "Do not forget your place, servant.", \
+				"I gave you all that power, and you still managed to fail me? Pathetic.", "I would normally be furious at you for this, but your continued inability to perform basic tasks that I had previously believed were idiot-proof quite honestly amuses me at this point.", "You'll be back.", "If you cannot scour this poison, I shall scour your meager life, insect!", "Just have a cultist use a ritual dagger on you, you imbecile!", "I gave you magic, friends, a PURPOSE! What have THEY ever done for you?!")].</span>")
```
Comment (coiax) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:274]: Should we put these strings somewhere else? Or at least separate them by line breaks?

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 6]
```diff
@@ -218,54 +218,73 @@
...
+	
+	if(current_cycle >= 25 && current_cycle < 60)		// ~10u or more of holy water is required to reach this point
 		if(!M.stuttering)
 			M.stuttering = 1
 		M.stuttering = min(M.stuttering+4, 10)
 		M.Dizzy(5)
 		if(iscultist(M) && prob(20))
-			M.say(pick("Av'te Nar'Sie","Pa'lid Mors","INO INO ORA ANA","SAT ANA!","Daim'niodeis Arc'iai Le'eones","R'ge Na'sie","Diabo us Vo'iscum","Eld' Mon Nobis"), forced = "holy water")
-			if(prob(10))
+			M.say(pick("Av'te Nar'Sie","Pa'lid Mors","INO INO ORA ANA","SAT ANA!","Daim'niodeis Arc'iai Le'eones","R'ge Nar'sie","Diabo us Vo'iscum","Eld' Mon Nobis"), forced = "holy water")
```
Comment (coiax) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:266]:
```suggestion
			M.say(pick("Av'te Nar'Sie","Pa'lid Mors","INO INO ORA ANA","SAT ANA!","Daim'niodeis Arc'iai Le'eones","R'ge Nar'sie","Diabo us Vo'iscum","Eld' Mon Nobis"), forced = type)
```

File: code/modules/reagents/chemistry/reagents/other_reagents.dm (hunk 2 above)
Comment (coiax) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:242]:
```suggestion
			exposed_mob.say("AAAHHH! YOU CURSED BRAT! LOOK WHAT YOU'VE DONE! I'm MELTING! Melting! Ohhh, what a world, what a world! Who would have thought a good little spaceman like you could destroy my beautiful wickedness?! AUUUUGH! I'm gone! I'm gone! I'm going... Augh... Aaauuggghhh...", forced = type) //you can theoretically say this and live using something that lets you avoid falling into crit, but eh, I'm not gonna bother to add a bunch of code for those edge cases for such a niche occurrence
```

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 7]
```diff
@@ -218,54 +218,73 @@
 	glass_icon_state  = "glass_clear"
 	glass_name = "glass of holy water"
 	glass_desc = "A glass of holy water."
+	metabolization_rate = 0.4 //a fixed consumption rate to prevent balancing going out of whack
 	self_consuming = TRUE //divine intervention won't be limited by the lack of a liver
+	var/seizure = FALSE //has Nar'Sie chewed you out during this deconversion process yet?
```
Comment (bobbahbrown) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:223]: This isn't dmdoc compliant, please follow our [code contribution guidelines](https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md#documenting-a-variabledefine) regarding class-level variable declarations and their documentation.

File: code/modules/reagents/chemistry/reagents/other_reagents.dm (hunk 2 above)
Comment (bobbahbrown) [reply code/modules/reagents/chemistry/reagents/other_reagents.dm:242]: i agree that this is seems excessive but maybe with 70 units its not likely to happen

Comment (ATH1909) [reply code/modules/reagents/chemistry/reagents/other_reagents.dm]: > What would it take for you to get a reac volume of 70u though?

it has to be a reac_volume of 70u in one go, so you'd have to splash a bucket's worth of more of holy water on them
